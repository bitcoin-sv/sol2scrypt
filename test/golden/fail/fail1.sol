pragma solidity ^0.8.10;

interface ICounter {
    function count() external view returns (uint);

    function increment() external;
}

contract A {
    function foo() public pure virtual returns (string memory) {
        return "A";
    }
}

// Contracts inherit other contracts by using the keyword 'is'.
contract B is A {
    // Override A.foo()
    function foo() public pure virtual override returns (string memory) {
        return "B";
    }
}

// External contract used for try / catch examples
contract Foo {
    address public owner;

    enum Status {
        Pending,
        Shipped,
        Accepted,
        Rejected,
        Canceled
    }

    struct Todo {
        string text;
        bool completed;
    }
    Todo[] public todos;

    constructor(address _owner) {
        require(_owner != address(0), "invalid address");
        owner = _owner;
    }

    modifier onlyOwner() {
        require(msg.sender == owner, "Not owner");
        // Underscore is a special character only used inside
        // a function modifier and it tells Solidity to
        // execute the rest of the code.
        _;
    }

    function myFunc(uint x, uint y) public pure returns (string memory) {
        require(x != 0, "require failed");
        return "my func was called";
    }

    function myFunc(uint x) public pure returns (string memory) {
        myFunc(1,1);
    }

    function create(string memory _text) public {
        // 3 ways to initialize a struct
        // - calling it like a function
        todos.push(Todo(_text, false));

        // key value mapping
        todos.push(Todo({text: _text, completed: false}));
        
        uint len = todos.length;
    }
}



contract SimpleStorage {
    uint storedData;
    uint public i = 0;
    mapping(uint => address) private _owners;
    function set(uint x) external {
        bytes memory a = msg.data;

        uint aa = 1 seconds;
        int min = type(int).min;
        uint c = block.number;

        address d = tx.origin;

        addmod(4, 5, 3);

        mulmod(4, 5, 3);

        (x, x) = (1, 3);

        (uint8 p1, uint8 p2) = (1, 3);

        Foo foo = new Foo(msg.sender);

        storedData = x;
    }

    function get() public view returns (uint) {
        keccak256(abi.encodePacked("aaa"));
        return storedData;
    }

    function testCallFoo(address payable _addr) public payable {
        // You can send ether and specify a custom gas amount
        // (bool success, bytes memory data) = _addr.call{value: msg.value, gas: 5000}(
        //     abi.encodeWithSignature("foo(string,uint256)", "call foo", 123)
        // );
    }


    function ownerOf(uint tokenId) internal view  returns (address owner) {
        owner = _owners[tokenId];
        require(owner != address(0), "token doesn't exist");
    }
}