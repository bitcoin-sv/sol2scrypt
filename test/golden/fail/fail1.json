[
    {
        "src": "test/golden/fail/fail1.sol:3:1:7:2",
        "type": "Error",
        "message": "unsupported contract definition `ContractDefinition`"
    },
    {
        "src": "test/golden/fail/fail1.sol:16:1:21:2",
        "type": "Error",
        "message": "unsupported contract definition `ContractDefinition`"
    },
    {
        "src": "test/golden/fail/fail1.sol:27:6:33:6",
        "type": "Error",
        "message": "unsupported contract part `ContractPartEnumDefinition`"
    },
    {
        "src": "test/golden/fail/fail1.sol:39:5:39:11",
        "type": "Error",
        "message": "array length should be explicitly specified"
    },
    {
        "src": "test/golden/fail/fail1.sol:46:5:52:6",
        "type": "Error",
        "message": "unsupported contract part `ContractPartModifierDefinition`"
    },
    {
        "src": "test/golden/fail/fail1.sol:59:5:61:6",
        "type": "Error",
        "message": "duplicate function name `myFunc` in contract"
    },
    {
        "src": "test/golden/fail/fail1.sol:79:5:79:23",
        "type": "Error",
        "message": "unsupported state variable with init value"
    },
    {
        "src": "test/golden/fail/fail1.sol:81:26:81:34",
        "type": "Error",
        "message": "unsupported expression: `msg.data`"
    },
    {
        "src": "test/golden/fail/fail1.sol:83:19:83:28",
        "type": "Error",
        "message": "unsupported expression : `Literal`"
    },
    {
        "src": "test/golden/fail/fail1.sol:84:19:84:28",
        "type": "Error",
        "message": "unsupported function call : `type`"
    },
    {
        "src": "test/golden/fail/fail1.sol:85:18:85:30",
        "type": "Error",
        "message": "unsupported expression: `block.number`"
    },
    {
        "src": "test/golden/fail/fail1.sol:87:21:87:30",
        "type": "Error",
        "message": "unsupported expression: `tx.origin`"
    },
    {
        "src": "test/golden/fail/fail1.sol:89:9:89:24",
        "type": "Error",
        "message": "unsupported function call : `addmod`"
    },
    {
        "src": "test/golden/fail/fail1.sol:91:9:91:24",
        "type": "Error",
        "message": "unsupported function call : `mulmod`"
    },
    {
        "src": "test/golden/fail/fail1.sol:93:9:93:25",
        "type": "Error",
        "message": "unsupported SimpleStatementVariableAssignmentList"
    },
    {
        "src": "test/golden/fail/fail1.sol:95:19:95:26",
        "type": "Error",
        "message": "unsupported expression : `New`"
    },
    {
        "src": "test/golden/fail/fail1.sol:95:9:95:12",
        "type": "Error",
        "message": "unsupported type `TypeNameUserDefinedTypeName`"
    },
    {
        "src": "test/golden/fail/fail1.sol:95:9:95:39",
        "type": "Error",
        "message": "unsupported SimpleStatementVariableDeclarationList"
    },
    {
        "src": "test/golden/fail/fail1.sol:98:9:100:10",
        "type": "Error",
        "message": "unsupported statement `WhileStatement`"
    },
    {
        "src": "test/golden/fail/fail1.sol:102:9:111:10",
        "type": "Error",
        "message": "unsupported statement `ForStatement`"
    },
    {
        "src": "test/golden/fail/fail1.sol:118:24:118:26",
        "type": "Error",
        "message": "unsupported binary operator `**`"
    },
    {
        "src": "test/golden/fail/fail1.sol:119:9:119:43",
        "type": "Error",
        "message": "unsupported function call : `keccak256`"
    },
    {
        "src": "test/golden/fail/fail1.sol:123:26:123:41",
        "type": "Error",
        "message": "unsupported type `TypeNameElementaryTypeName`"
    }
]